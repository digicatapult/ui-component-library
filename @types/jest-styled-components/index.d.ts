declare module 'jest-styled-components/serializer'
