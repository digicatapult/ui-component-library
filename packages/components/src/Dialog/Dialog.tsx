import React, {useCallback, useEffect, useRef, useState} from 'react'
import styled from 'styled-components'
import Button, {ButtonPrimary, ButtonDanger, ButtonProps} from '../deprecated/Button'
import Box from '../Box'
import {get} from '../constants'
import {useOnEscapePress, useProvidedRefOrCreate} from '../hooks'
import {useFocusTrap} from '../hooks/useFocusTrap'
import sx, {SxProp} from '../sx'
import StyledOcticon from '../StyledOcticon'
import {XIcon} from '@primer/octicons-react'
import {useFocusZone} from '../hooks/useFocusZone'
import {FocusKeys} from '@primer/behaviors'
import Portal from '../Portal'
import {useRefObjectAsForwardedRef} from '../hooks/useRefObjectAsForwardedRef'
import {useSSRSafeId} from '@react-aria/ssr'

const ANIMATION_DURATION = '200ms'

/**
 * Props that characterize a button to be rendered into the footer of
 * a Dialog.
 */
export type DialogButtonProps = ButtonProps & {
  /**
   * The type of Button element to use
   */
  buttonType?: 'normal' | 'primary' | 'danger'

  /**
   * The Button's inner text
   */
  content: React.ReactNode

  /**
   * If true, and if this is the only button with autoFocus set to true,
   * focus this button automatically when the dialog appears.
   */
  autoFocus?: boolean

  /**
   * A reference to the rendered Button’s DOM node, used together with
   * `autoFocus` for `focusTrap`’s `initialFocus`.
   */
  ref?: React.RefObject<HTMLButtonElement>
}

/**
 * Props to customize the rendering of the Dialog.
 */
export interface DialogProps {
  /**
   * Title of the Dialog. Also serves as the aria-label for this Dialog.
   */
  title?: React.ReactNode

  /**
   * The Dialog's subtitle. Optional. Rendered below the title in smaller
   * type with less contrast. Also serves as the aria-describedby for this
   * Dialog.
   */
  subtitle?: React.ReactNode

  /**
   * Provide a custom renderer for the dialog header. This content is
   * rendered directly into the dialog body area, full bleed from edge
   * to edge, top to the start of the body element.
   *
   * Warning: using a custom renderer may violate Primer UX principles.
   */
  renderHeader?: React.FunctionComponent<React.PropsWithChildren<DialogHeaderProps>>

  /**
   * Provide a custom render function for the dialog body. This content is
   * rendered directly into the dialog body area, full bleed from edge to
   * edge, header to footer.
   *
   * Warning: using a custom renderer may violate Primer UX principles.
   */
  renderBody?: React.FunctionComponent<React.PropsWithChildren<DialogProps>>

  /**
   * Provide a custom render function for the dialog footer. This content is
   * rendered directly into the dialog footer area, full bleed from edge to
   * edge, end of the body element to bottom.
   *
   * Warning: using a custom renderer may violate Primer UX principles.
   */
  renderFooter?: React.FunctionComponent<React.PropsWithChildren<DialogProps>>

  /**
   * Specifies the buttons to be rendered in the Dialog footer.
   */
  footerButtons?: DialogButtonProps[]

  /**
   * This method is invoked when a gesture to close the dialog is used (either
   * an Escape key press or clicking the "X" in the top-right corner). The
   * gesture argument indicates the gesture that was used to close the dialog
   * (either 'close-button' or 'escape').
   */
  onClose: (gesture: 'close-button' | 'escape') => void

  /**
   * Default: "dialog". The ARIA role to assign to this dialog.
   * @see https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal
   * @see https://www.w3.org/TR/wai-aria-practices-1.1/#alertdialog
   */
  role?: 'dialog' | 'alertdialog'

  /**
   * The width of the dialog.
   * small: 296px
   * medium: 320px
   * large: 480px
   * xlarge: 640px
   */
  width?: DialogWidth

  /**
   * The height of the dialog.
   * small: 296x480
   * large: 480x640
   * auto: variable based on contents
   */
  height?: DialogHeight
}

/**
 * Props that are passed to a component that serves as a dialog header
 */
export interface DialogHeaderProps extends DialogProps {
  /**
   * ID of the element that will be used as the `aria-labelledby` attribute on the
   * dialog. This ID should be set to the element that renders the dialog's title.
   */
  dialogLabelId: string

  /**
   * ID of the element that will be used as the `aria-describedby` attribute on the
   * dialog. This ID should be set to the element that renders the dialog's subtitle.
   */
  dialogDescriptionId: string
}

const Backdrop = styled('div')`
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.4);
  animation: dialog-backdrop-appear ${ANIMATION_DURATION} ${get('animation.easeOutCubic')};

  @keyframes dialog-backdrop-appear {
    0% {
      opacity: 0;
    }
    100% {
      opacity: 1;
    }
  }
`

const heightMap = {
  small: '480px',
  large: '640px',
  auto: 'auto',
} as const

const widthMap = {
  small: '296px',
  medium: '320px',
  large: '480px',
  xlarge: '640px',
} as const

export type DialogWidth = keyof typeof widthMap
export type DialogHeight = keyof typeof heightMap

type StyledDialogProps = {
  width?: DialogWidth
  height?: DialogHeight
} & SxProp

const StyledDialog = styled.div<StyledDialogProps>`
  display: flex;
  flex-direction: column;
  background-color: ${get('colors.canvas.overlay')};
  box-shadow: ${get('shadows.overlay.shadow')};
  min-width: 296px;
  max-width: calc(100vw - 64px);
  max-height: calc(100vh - 64px);
  width: ${props => widthMap[props.width ?? ('xlarge' as const)]};
  height: ${props => heightMap[props.height ?? ('auto' as const)]};
  border-radius: 12px;
  opacity: 1;
  animation: overlay--dialog-appear ${ANIMATION_DURATION} ${get('animation.easeOutCubic')};

  @keyframes overlay--dialog-appear {
    0% {
      opacity: 0;
      transform: scale(0.5);
    }
    100% {
      opacity: 1;
      transform: scale(1);
    }
  }

  ${sx};
`

const DefaultHeader: React.FC<React.PropsWithChildren<DialogHeaderProps>> = ({
  dialogLabelId,
  title,
  subtitle,
  dialogDescriptionId,
  onClose,
}) => {
  const onCloseClick = useCallback(() => {
    onClose('close-button')
  }, [onClose])
  return (
    <Dialog.Header>
      <Box display="flex">
        <Box display="flex" px={2} py="6px" flexDirection="column" flexGrow={1}>
          <Dialog.Title id={dialogLabelId}>{title ?? 'Dialog'}</Dialog.Title>
          {subtitle && <Dialog.Subtitle id={dialogDescriptionId}>{subtitle}</Dialog.Subtitle>}
        </Box>
        <Dialog.CloseButton onClose={onCloseClick} />
      </Box>
    </Dialog.Header>
  )
}
const DefaultBody: React.FC<React.PropsWithChildren<DialogProps>> = ({children}) => {
  return <Dialog.Body>{children}</Dialog.Body>
}
const DefaultFooter: React.FC<React.PropsWithChildren<DialogProps>> = ({footerButtons}) => {
  const {containerRef: footerRef} = useFocusZone({
    bindKeys: FocusKeys.ArrowHorizontal | FocusKeys.Tab,
    focusInStrategy: 'closest',
  })
  return footerButtons ? (
    <Dialog.Footer ref={footerRef as React.RefObject<HTMLDivElement>}>
      <Dialog.Buttons buttons={footerButtons} />
    </Dialog.Footer>
  ) : null
}

const _Dialog = React.forwardRef<HTMLDivElement, React.PropsWithChildren<DialogProps>>((props, forwardedRef) => {
  const {
    title = 'Dialog',
    subtitle = '',
    renderHeader,
    renderBody,
    renderFooter,
    onClose,
    role = 'dialog',
    width = 'xlarge',
    height = 'auto',
    footerButtons = [],
  } = props
  const dialogLabelId = useSSRSafeId()
  const dialogDescriptionId = useSSRSafeId()
  const autoFocusedFooterButtonRef = useRef<HTMLButtonElement>(null)
  for (const footerButton of footerButtons) {
    if (footerButton.autoFocus) {
      footerButton.ref = autoFocusedFooterButtonRef
    }
  }
  const defaultedProps = {...props, title, subtitle, role, dialogLabelId, dialogDescriptionId}

  const dialogRef = useRef<HTMLDivElement>(null)
  useRefObjectAsForwardedRef(forwardedRef, dialogRef)
  const backdropRef = useRef<HTMLDivElement>(null)
  useFocusTrap({containerRef: dialogRef, restoreFocusOnCleanUp: true, initialFocusRef: autoFocusedFooterButtonRef})

  useOnEscapePress(
    (event: KeyboardEvent) => {
      onClose('escape')
      event.preventDefault()
    },
    [onClose],
  )

  const header = (renderHeader ?? DefaultHeader)(defaultedProps)
  const body = (renderBody ?? DefaultBody)(defaultedProps)
  const footer = (renderFooter ?? DefaultFooter)(defaultedProps)

  return (
    <>
      <Portal>
        <Backdrop ref={backdropRef}>
          <StyledDialog
            width={width}
            height={height}
            ref={dialogRef}
            role={role}
            aria-labelledby={dialogLabelId}
            aria-describedby={dialogDescriptionId}
          >
            {header}
            {body}
            {footer}
          </StyledDialog>
        </Backdrop>
      </Portal>
    </>
  )
})
_Dialog.displayName = 'Dialog'

const Header = styled.div<SxProp>`
  box-shadow: 0 1px 0 ${get('colors.border.default')};
  padding: ${get('space.2')};
  z-index: 1;
  flex-shrink: 0;
`

const Title = styled.h1<SxProp>`
  font-size: ${get('fontSizes.1')};
  font-weight: ${get('fontWeights.bold')};
  margin: 0; /* override default margin */
  ${sx};
`

const Subtitle = styled.h2<SxProp>`
  font-size: ${get('fontSizes.0')};
  color: ${get('colors.fg.muted')};
  margin: 0; /* override default margin */
  margin-top: ${get('space.1')};

  ${sx};
`

const Body = styled.div<SxProp>`
  flex-grow: 1;
  overflow: auto;
  padding: ${get('space.3')};

  ${sx};
`

const Footer = styled.div<SxProp>`
  box-shadow: 0 -1px 0 ${get('colors.border.default')};
  padding: ${get('space.3')};
  display: flex;
  flex-flow: wrap;
  justify-content: flex-end;
  z-index: 1;
  flex-shrink: 0;

  button {
    margin-left: ${get('space.1')};
    &:first-child {
      margin-left: 0;
    }
  }

  ${sx};
`

const buttonTypes = {
  normal: Button,
  primary: ButtonPrimary,
  danger: ButtonDanger,
}
const Buttons: React.FC<React.PropsWithChildren<{buttons: DialogButtonProps[]}>> = ({buttons}) => {
  const autoFocusRef = useProvidedRefOrCreate<HTMLButtonElement>(buttons.find(button => button.autoFocus)?.ref)
  let autoFocusCount = 0
  const [hasRendered, setHasRendered] = useState(0)
  useEffect(() => {
    // hack to work around dialogs originating from other focus traps.
    if (hasRendered === 1) {
      autoFocusRef.current?.focus()
    } else {
      setHasRendered(hasRendered + 1)
    }
  }, [autoFocusRef, hasRendered])

  return (
    <>
      {buttons.map((dialogButtonProps, index) => {
        const {content, buttonType = 'normal', autoFocus = false, ...buttonProps} = dialogButtonProps
        const ButtonElement = buttonTypes[buttonType]
        return (
          <ButtonElement
            key={index}
            {...buttonProps}
            variant={buttonType}
            ref={autoFocus && autoFocusCount === 0 ? (autoFocusCount++, autoFocusRef) : null}
          >
            {content}
          </ButtonElement>
        )
      })}
    </>
  )
}
const DialogCloseButton = styled(Button)`
  border-radius: 4px;
  background: transparent;
  border: 0;
  vertical-align: middle;
  color: ${get('colors.fg.muted')};
  padding: ${get('space.2')};
  align-self: flex-start;
  line-height: normal;
  box-shadow: none;
`
const CloseButton: React.FC<React.PropsWithChildren<{onClose: () => void}>> = ({onClose}) => {
  return (
    <DialogCloseButton aria-label="Close" onClick={onClose}>
      <StyledOcticon icon={XIcon} />
    </DialogCloseButton>
  )
}

/**
 * A dialog is a type of overlay that can be used for confirming actions, asking
 * for disambiguation, and presenting small forms. They generally allow the user
 * to focus on a quick task without having to navigate to a different page.
 *
 * Dialogs appear in the page after a direct user interaction. Don't show dialogs
 * on page load or as system alerts.
 *
 * Dialogs appear centered in the page, with a visible backdrop that dims the rest
 * of the window for focus.
 *
 * All dialogs have a title and a close button.
 *
 * Dialogs are modal. Dialogs can be dismissed by clicking on the close button,
 * pressing the escape key, or by interacting with another button in the dialog.
 * To avoid losing information and missing important messages, clicking outside
 * of the dialog will not close it.
 *
 * The sub components provided (e.g. Header, Title, etc.) are available for custom
 * renderers only. They are not intended to be used otherwise.
 */
export const Dialog = Object.assign(_Dialog, {
  Header,
  Title,
  Subtitle,
  Body,
  Footer,
  Buttons,
  CloseButton,
})
