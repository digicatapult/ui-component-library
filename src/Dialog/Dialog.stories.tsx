import React, { PropsWithChildren, useEffect, useRef, useState } from 'react'
import { Story } from '@storybook/react'
import styled from 'styled-components'

import { Grid, Section } from '../'
import { default as DialogComponent } from './index'

export default {
  title: 'Components/Dialog',
  component: DialogComponent,
  argTypes: {
    includeClose: {
      control: {
        type: 'boolean',
      },
    },
  },
}

const DialogStoryTemplate: Story = (args) => {
  return <DefaultStoryHost {...args} useModal={false} />
}
export const Dialog = DialogStoryTemplate.bind({})
Dialog.args = {
  includeClose: true,
  modalBackdropColor: 'rgba(255, 0, 0, 0.05)',
}

const ModalStoryTemplate: Story = (args) => {
  return <DefaultStoryHost {...args} useModal={true} />
}
export const Modal = ModalStoryTemplate.bind({})
Modal.args = {
  includeClose: true,
  modalBackdropColor: 'rgba(255, 0, 0, 0.05)',
}

const DefaultStoryHost: React.FC<any> = ({ useModal, ...args }) => {
  const dialogRef = useRef<HTMLDialogElement>(null)
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const listener = () => {
      setIsOpen(false)
    }
    const dialog = dialogRef.current
    dialog?.addEventListener('close', listener)

    return () => dialog?.removeEventListener('close', listener)
  })

  return (
    <div>
      <button
        onClick={() => {
          setIsOpen(!isOpen)
          if (!isOpen) {
            if (useModal) dialogRef.current?.showModal()
            else dialogRef.current?.show()
          } else dialogRef.current?.close()
        }}
      >
        {!isOpen ? 'Open' : 'Close'}
      </button>
      <DialogComponent ref={dialogRef} {...args}>
        <MockContent />
      </DialogComponent>
    </div>
  )
}

const MockContent = styled.div`
  margin: 2ch;
  background: cyan;
  width: 300px;
  height: 300px;
`

export const HII: Story = () => {
  const dialogRef = useRef<HTMLDialogElement>(null)
  const [isOpen, setIsOpen] = useState(false)

  useEffect(() => {
    const listener = () => {
      setIsOpen(false)
    }
    const dialog = dialogRef.current
    dialog?.addEventListener('close', listener)

    return () => dialog?.removeEventListener('close', listener)
  })

  return (
    <div style={{
      height:"600px",
      background: '#e0e0f0'
    }}>
      <button
        onClick={() => {
          setIsOpen(!isOpen)
          if (!isOpen) {
            dialogRef.current?.show()
          } else {
            dialogRef.current?.close()
          }
        }}
      >
        {!isOpen ? 'Open' : 'Close'}
      </button>
      <DialogComponent ref={dialogRef} includeClose={true}>
        <Section headingLevel={2} title="Name" padding="1em 1.5em" width='47ch' headingSize="2em" background='white'>
          <Grid areas={[]} rows={["auto"]} columns={["auto"]} gap="1em">
            <span>Name of Lead Partner</span>
            <Section headingLevel={3} title="Project details" padding='1em 1.5em' margin="0 -1.5em" background='#dfe667' headingSize="1em" >
              Scope / goal - The £33 million Low Carbon Hydrogen Supply competition aimed to accelerate the development of low carbon bulk hydrogen supply solutions in specific sectors. It was aimed at projects at a TRL of 4 to 7, which could result in lower capital or operating costs when compared to Steam Methane Reformation with Carbon Capture & Storage (SMR+CCS), or improve the carbon capture rates at a comparable cost.
            </Section>
            <Section headingLevel={3} title="Funding" padding='0em 0' headingSize="1em" background='white'>
              £240m
            </Section>
          </Grid>
        </Section>
      </DialogComponent>
    </div>
  )
}
