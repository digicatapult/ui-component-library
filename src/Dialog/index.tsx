import React, { PropsWithChildren, useEffect, useRef } from 'react'
import styled from 'styled-components'

export interface DialogProps {
  includeClose?: boolean
  modalBackdropColor?: string

  borderRadius?: string
  padding?: string
  margin?: string
  border?: string
  boxShadow?: string
  width?: string
  height?: string
  maxHeight?: string
}

const Dialog = React.forwardRef<
  HTMLDialogElement,
  PropsWithChildren<DialogProps>
>(
  (
    {
      children,
      includeClose = true,
      borderRadius = '20px',
      padding = '0',
      margin = '0 auto',
      border = '0',
      boxShadow = '5px 5px 5px #90909090',
      width = 'fit-content',
      height,
      modalBackdropColor = 'unset',
      maxHeight,
    },
    dialogRef
  ) => {
    return (
      <Wrapper
        ref={dialogRef}
        modalBackdropColor={modalBackdropColor}
        borderRadius={borderRadius}
        padding={padding}
        margin={margin}
        border={border}
        boxShadow={boxShadow}
        width={width}
        height={height}
        maxHeight={maxHeight}
      >
        <DialogForm method="dialog">
          {includeClose ? <CloseButton /> : null}
          {children}
        </DialogForm>
      </Wrapper>
    )
  }
)

const Wrapper = styled.dialog<DialogProps>`
  position: absolute;
  overflow: hidden;

  border-radius: ${({ borderRadius }) => borderRadius};
  padding: ${({ padding }) => padding};
  margin: ${({ margin }) => margin};
  border: ${({ border }) => border};
  box-shadow: ${({ boxShadow }) => boxShadow};
  width: ${({ width }) => width};
  height: ${({ height }) => height};
  max-height: ${({ maxHeight }) => maxHeight};

  ::backdrop {
    background: ${({ modalBackdropColor }) => modalBackdropColor};
  }
`

const DialogForm = styled.form`
  display: inline-block;
  position: relative;
  overflow: auto;
  max-height: inherit;
  height: inherit;
`

const CloseButton = styled.button`
  position: absolute;
  top: 1em;
  right: 1em;
  width: 1.5em;
  height: 1.5em;
  border: 0;
  padding: 0;
  background: transparent;

  ::before,
  ::after {
    content: '';
    position: absolute;
    left: 0;
    height: 2px;
    width: calc(100% * 1.41);
    background: black;
  }

  ::before {
    transform: translateX(-0.3075em) rotateZ(45deg);
  }

  ::after {
    transform: translateX(-0.3075em) rotateZ(-45deg);
  }
`

export default Dialog
