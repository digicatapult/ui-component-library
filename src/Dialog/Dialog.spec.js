import React, { useRef, useState, useEffect } from 'react'
import renderer, { act } from 'react-test-renderer'
import styled from 'styled-components'
import 'jest-styled-components'

import Dialog from './index'

describe('Dialog', () => {
  test('default', () => {
    const instance = renderer.create(
      <Dialog>
        <MockContent />
      </Dialog>
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('no close', () => {
    const instance = renderer.create(
      <Dialog includeClose={false}>
        <MockContent />
      </Dialog>
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('style overrides', () => {
    const instance = renderer.create(
      <Dialog
        modalBackdropColor="rgba(0, 255, 0, 0.05)"
        borderRadius="10px"
        padding="1em"
        margin="1em"
        border="1px solid black"
        boxShadow="10px 10px 10px #90909090"
        width="500px"
        height="500px"
      >
        <MockContent />
      </Dialog>
    )
    const tree = instance.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

const MockContent = styled.div`
  margin: 2ch;
  background: cyan;
  width: 300px;
  height: 300px;
`
