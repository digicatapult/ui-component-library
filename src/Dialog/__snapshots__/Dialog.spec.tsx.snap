// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog default 1`] = `
.c0 {
  position: absolute;
  overflow: hidden;
  border-radius: 20px;
  padding: 0;
  margin: 0 auto;
  border: 0;
  box-shadow: 5px 5px 5px #90909090;
  width: fit-content;
}

.c0::backdrop {
  background: unset;
}

.c1 {
  display: inline-block;
  position: relative;
  overflow: auto;
  max-height: inherit;
  height: inherit;
}

.c2 {
  position: absolute;
  top: 1em;
  right: 1em;
  width: 1.5em;
  height: 1.5em;
  border: 0;
  padding: 0;
  background: transparent;
}

.c2:before,
.c2:after {
  content: '';
  position: absolute;
  left: 0;
  height: 2px;
  width: calc(100% * 1.41);
  background: black;
}

.c2:before {
  transform: translateX(-0.3075em) rotateZ(45deg);
}

.c2:after {
  transform: translateX(-0.3075em) rotateZ(-45deg);
}

.c3 {
  margin: 2ch;
  background: cyan;
  width: 300px;
  height: 300px;
}

<dialog
  border="0"
  borderRadius="20px"
  boxShadow="5px 5px 5px #90909090"
  className="c0"
  margin="0 auto"
  modalBackdropColor="unset"
  padding="0"
  width="fit-content"
>
  <form
    className="c1"
    method="dialog"
  >
    <button
      className="c2"
    />
    <div
      className="c3"
    />
  </form>
</dialog>
`;

exports[`Dialog no close 1`] = `
.c0 {
  position: absolute;
  overflow: hidden;
  border-radius: 20px;
  padding: 0;
  margin: 0 auto;
  border: 0;
  box-shadow: 5px 5px 5px #90909090;
  width: fit-content;
}

.c0::backdrop {
  background: unset;
}

.c1 {
  display: inline-block;
  position: relative;
  overflow: auto;
  max-height: inherit;
  height: inherit;
}

.c2 {
  margin: 2ch;
  background: cyan;
  width: 300px;
  height: 300px;
}

<dialog
  border="0"
  borderRadius="20px"
  boxShadow="5px 5px 5px #90909090"
  className="c0"
  margin="0 auto"
  modalBackdropColor="unset"
  padding="0"
  width="fit-content"
>
  <form
    className="c1"
    method="dialog"
  >
    <div
      className="c2"
    />
  </form>
</dialog>
`;

exports[`Dialog style overrides 1`] = `
.c0 {
  position: absolute;
  overflow: hidden;
  border-radius: 10px;
  padding: 1em;
  margin: 1em;
  border: 1px solid black;
  box-shadow: 10px 10px 10px #90909090;
  width: 500px;
  height: 500px;
}

.c0::backdrop {
  background: rgba(0, 255, 0, 0.05);
}

.c1 {
  display: inline-block;
  position: relative;
  overflow: auto;
  max-height: inherit;
  height: inherit;
}

.c2 {
  position: absolute;
  top: 1em;
  right: 1em;
  width: 1.5em;
  height: 1.5em;
  border: 0;
  padding: 0;
  background: transparent;
}

.c2:before,
.c2:after {
  content: '';
  position: absolute;
  left: 0;
  height: 2px;
  width: calc(100% * 1.41);
  background: black;
}

.c2:before {
  transform: translateX(-0.3075em) rotateZ(45deg);
}

.c2:after {
  transform: translateX(-0.3075em) rotateZ(-45deg);
}

.c3 {
  margin: 2ch;
  background: cyan;
  width: 300px;
  height: 300px;
}

<dialog
  border="1px solid black"
  borderRadius="10px"
  boxShadow="10px 10px 10px #90909090"
  className="c0"
  height="500px"
  margin="1em"
  modalBackdropColor="rgba(0, 255, 0, 0.05)"
  padding="1em"
  width="500px"
>
  <form
    className="c1"
    method="dialog"
  >
    <button
      className="c2"
    />
    <div
      className="c3"
    />
  </form>
</dialog>
`;
